worlds/tww/randomizers/Entrances.py

```diff
@@ -0,0 +1,866 @@
+                        if ex in terminal_exits and ex not in (DUNGEON_EXITS + BOSS_EXITS)
+                    ]
+
+            if not possible_remaining_exits:
+                raise FillError(f"No valid exits to place for entrance: {zone_entrance.entrance_name}")
+
+            zone_exit = self.multiworld.random.choice(possible_remaining_exits)
+            remaining_exits.remove(zone_exit)
+
+            self.entrance_connections[zone_entrance.entrance_name] = zone_exit.unique_name
+            self.done_entrances_to_exits[zone_entrance] = zone_exit
+            self.done_exits_to_entrances[zone_exit] = zone_entrance
+
+            if zone_exit in self.banned_exits:
+                # Keep track of which islands have a required bosses mode banned dungeon to avoid marker overlap.
+                if zone_exit in DUNGEON_EXITS + BOSS_EXITS:
+                    # We only keep track of dungeon exits and boss exits, not miniboss exits.
+                    # Banned miniboss exits can share an island with required dungeons/bosses.
+                    outer_entrance = self.get_outermost_entrance_for_entrance(zone_entrance)
+
+                    # Because we filter above so that we always assign entrances from the sea inwards, we can assume
+                    # that when we assign an entrance, it has a path back to the sea.
+                    # If we're assigning a non-terminal entrance, any nested entrances will get assigned after this one,
+                    # and we'll run through this code again (so we can reason based on `zone_exit` only instead of
+                    # having to recurse through the nested exits to find banned dungeons/bosses).
+                    assert outer_entrance and outer_entrance.island_name is not None
+                    self.islands_with_a_banned_dungeon.add(outer_entrance.island_name)
+
+    def finalize_all_randomized_sets_of_entrances(self) -> None:
+        """
+        Finalize all randomized entrance sets.
+
+        For all entrance-exit pairs, this function adds a connection with the appropriate access rule to the world.
+        """
+
+        def get_access_rule(region: str) -> str:
+            snake_case_region = region.lower().replace("'", "").replace(" ", "_")
+            return f"can_access_{snake_case_region}"
+
+        # Connect each entrance-exit pair in the multiworld with the access rule for the entrance.
+        for zone_entrance, zone_exit in self.done_entrances_to_exits.items():
+            entrance_region = self.world.get_region(zone_entrance.entrance_name)
+            exit_region = self.world.get_region(zone_exit.unique_name)
+            entrance_region.connect(
+                exit_region,
+                rule=lambda state, entrance=entrance_region.name: getattr(Macros, get_access_rule(entrance))(
+                    state, self.player
+                ),
+            )
+
+        if self.world.options.required_bosses:
```

Comment:
If the only purpose of this conditional branch is to run assertions, the condition can check `__debug__` to only run when assertions are enabled.

Maybe it could still be worth always running if calling `get_outermost_entrance_for_exit()` or `get_outermost_entrance_for_exit()` might pick up errors even when assertions are not enabled.

```suggestion
        if __debug__ and self.world.options.required_bosses:
```